File: 2125-number-of-laser-beams-in-a-bank/2125-number-of-laser-beams-in-a-bank.java
class Solution {

    private static int countOnes(String s) {
        int result = 0;
        for (int i = 0; i < s.length(); i++) if (s.charAt(i) == '1') result++;
        return result;
    }

    public int numberOfBeams(String[] bank) {
        int result = 0;

        int i = 0;
        while (i < bank.length) {
            if (i == bank.length - 1) {
                break;
            }

            String base = bank[i];
            int baseOnes = countOnes(base);
            if (baseOnes == 0) {
                i++;
                continue;
            }

            boolean isChanged = false;
            for (int j = i + 1; j < bank.length; j++) {
                String cmp = bank[j];
                int cmpOnes = countOnes(cmp);
                if (cmpOnes == 0) {
                    continue;
                }

                result += baseOnes * cmpOnes;
                i = j;
                isChanged = true;
                break;
            }

            if (!isChanged) {
                i++;
            }
        }


        return result;
    }
}

File: 2125-number-of-laser-beams-in-a-bank/2125-number-of-laser-beams-in-a-bank.swift
class Solution {
    private func countOnes(_ str: String) -> Int {
        var result = 0
        for char in str {
            if char == "1" {
                result += 1
            }
        }
        return result
    }
    
    func numberOfBeams(_ bank: [String]) -> Int {
        if bank.count == 1 {
            return 0
        }
        
        var result = 0
        var index = 0
        
        while index < bank.count - 1 {
            let lasers = bank[index]
            let lOnes = countOnes(lasers)
            if lOnes > 0 {
                var fIndex = -1
                var fOnes = -1
                for i in stride(from: index + 1, to: bank.count, by: 1) {
                    let frontLasers = bank[i]
                    let rOnes = countOnes(frontLasers)
                    if rOnes > 0 {
                        fOnes = rOnes
                        fIndex = i
                        break
                    }
                }
                if fIndex != -1 {
                    result += lOnes * fOnes
                    index = fIndex
                    continue
                }
                
            }
            index += 1
        }
        return result
    }
}
